File: Pascal's Triangle.cpp
class Solution {
public:
    vector<vector<int> > generate(int numRows) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<int> pre(1, 1);
        vector<vector<int> > result;
        
        for(int i = 0; i< numRows; i++)
        {
            vector<int> row(i+1);
            for(int j = 1; j < i; j++)
            {
                row[j] = pre[j-1] + pre[j];
            }
            row[0] = 1;
            row[i] = 1; 
            result.push_back(row);
            pre = row;
        }
        
        return result;
        
    }
};

File: Pascal's Triangle.java
public class Solution {
    public ArrayList<ArrayList<Integer>> generate(int numRows) {
        // Start typing your Java solution below
        // DO NOT write main() function
        
        //for dynamic programming if just use one demension array, please use back order so that new result won't depend on new result
        ArrayList<ArrayList<Integer>>ret = new ArrayList<ArrayList<Integer>>();  
        ArrayList<Integer> row = new ArrayList<Integer>();
        for(int i = 0; i < numRows; i++)
        {
            row.add(1);
            for(int j = i-1; j > 0; j--)
            {
                int val = row.get(j-1)+row.get(j);
                row.set(j, val);
            }
            ret.add((ArrayList<Integer>)row.clone());
        }
        
        return ret;
    }
}